### PR TITLE
Introduce HTTP-bridge & migrate Shelly Plug S

### PR DESCRIPTION
**Pull request by sebastianasen**
Introduce a new bridge for http connections and communication. 

Currently, each implementation of a device via RESTful APIs is implemented in its own bundle and is more or less stable depending on asynchronous calls.

In this PR it is currently used by the shelly IO plug as a first test. 

Co-authored-by: Michael Grill  <59126309+michaelgrill@users.noreply.github.com>

### CODE REVIEW

File: io.openems.edge.io.shelly/src/io/openems/edge/io/shelly/shellyplug/IoShellyPlugImpl.java [hunk 1]
```diff
@@ -64,9 +70,21 @@ public IoShellyPlugImpl() {
...
+
+		if (!this.isEnabled()) {
+			return;
+		}
+		this.httpBridge.subscribeJsonEveryCycle(this.baseUrl + "/status", t -> {
+			this.processHttpResult(t);
+			this._setSlaveCommunicationFailed(false);
+		}, t -> {
+			this.logDebug(this.log, t.getMessage());
+			this._setRelay(null);
```
Comment (sfeilmeier) [inline io.openems.edge.io.shelly/src/io/openems/edge/io/shelly/shellyplug/IoShellyPlugImpl.java:85]: The problem here is, that the 'Relay' Channel gets set properly, but you forgot `ActivePower` and `ActiveProductionEnergy`. This is a very typical mistake, but will lead to wrong data, e.g. in the UI.

My approach would be to use one common `BiConsumer` that handles the result and a exception. Also `processHttpResult` could then take both and act accordingly. This way you set all Channels in one place and it's more likely to not forget anything.

**Review by sfeilmeier — CHANGES_REQUESTED** [review]

File: io.openems.edge.bridge.http/src/io/openems/edge/bridge/http/BridgeHttpImpl.java [hunk 2]
```diff
@@ -0,0 +1,157 @@
...
+		switch (event.getTopic()) {
+		// TODO: Execute before TOPIC_CYCLE_BEFORE_PROCESS_IMAGE, like modbus bridge
+		case EdgeEventConstants.TOPIC_CYCLE_BEFORE_PROCESS_IMAGE -> {
+
+			this.endpoints.forEach(EndpointCountdown::decreaseCycleCount);
+
+			if (this.endpoints.isEmpty()) {
+				return;
+			}
+			while (this.endpoints.peek().getCycleCount() == 0) {
```
[the remark this answers is not in the document]
Comment (michaelgrill) [reply io.openems.edge.bridge.http/src/io/openems/edge/bridge/http/BridgeHttpImpl.java:121]: i also prefer that way in general but what if the items `getCycleCount()` is not `0` then i would have to add it again

File: io.openems.edge.bridge.http/src/io/openems/edge/bridge/http/BridgeHttpImpl.java [hunk 3]
```diff
@@ -0,0 +1,157 @@
...
+		this.connectTimeout = connectTimeout;
+		this.readTimeout = readTimeout;
+	}
+
+	private void handleEvent(Event event) {
+		switch (event.getTopic()) {
+		// TODO: Execute before TOPIC_CYCLE_BEFORE_PROCESS_IMAGE, like modbus bridge
+		case EdgeEventConstants.TOPIC_CYCLE_BEFORE_PROCESS_IMAGE -> {
+
+			this.endpoints.forEach(EndpointCountdown::decreaseCycleCount);
```
[the remark this answers is not in the document]
Comment (michaelgrill) [reply io.openems.edge.bridge.http/src/io/openems/edge/bridge/http/BridgeHttpImpl.java:116]: no the loop is only to handle endpoints wich have hit `cycleCount` 0 before that i decrease every cycle

**Review by sfeilmeier — APPROVED** [review]

**Comment by Sn0w3y** [comment]
```
2024-02-06T06:01:56,912 [tchQueue] ERROR [Events.Framework              ] FrameworkEvent ERROR
org.osgi.framework.ServiceException: Service factory returned null. (Component: IO.Shelly.Plus3EM (113))
	at org.apache.felix.framework.ServiceRegistrationImpl.getFactoryUnchecked(ServiceRegistrationImpl.java:385) ~[?:?]
	at org.apache.felix.framework.ServiceRegistrationImpl.getService(ServiceRegistrationImpl.java:249) ~[?:?]
	at org.apache.felix.framework.ServiceRegistry.getService(ServiceRegistry.java:362) ~[?:?]
	at org.apache.felix.framework.Felix.getService(Felix.java:3984) ~[?:?]
	at org.apache.felix.framework.BundleContextImpl.getService(BundleContextImpl.java:450) ~[?:?]
	at org.apache.felix.eventadmin.impl.handler.EventHandlerProxy.obtain(EventHandlerProxy.java:296) ~[?:?]
	at org.apache.felix.eventadmin.impl.handler.EventHandlerProxy.sendEvent(EventHandlerProxy.java:424) ~[?:?]
	at org.apache.felix.eventadmin.impl.tasks.HandlerTask.runWithoutDenylistTiming(HandlerTask.java:82) ~[?:?]
	at org.apache.felix.eventadmin.impl.tasks.SyncDeliverTasks.execute(SyncDeliverTasks.java:107) ~[?:?]
	at org.apache.felix.eventadmin.impl.handler.EventAdminImpl.sendEvent(EventAdminImpl.java:155) ~[?:?]
	at org.apache.felix.eventadmin.impl.security.EventAdminSecurityDecorator.sendEvent(EventAdminSecurityDecorator.java:96) ~[?:?]
	at io.openems.common.event.EventBuilder.send(EventBuilder.java:68) ~[?:?]
	at io.openems.edge.core.cycle.CycleWorker.forever(CycleWorker.java:159) ~[?:?]
	at io.openems.common.worker.AbstractWorker$1.run(AbstractWorker.java:156) ~[?:?]
```
	
	
	Happens after one deactivates for example the Shelly Device

**Pull request merged**